### PR TITLE
MySQL sessions + Prevent multiple sessions

### LINKED ISSUE #138: [Feature Request] Limit sessions per user

### PR DESCRIPTION
**Pull request by versx**
Saves sessions to mysql database instead of cookie based sessions.
Configurable option to clear existing sessions for a user if there are multiple
Fix #138

### CODE REVIEW

**Comment by versx** [comment]
Someone test please

**Comment by sketchysocks** [comment]
```
Failed to connect to the database: undefined
(node:44889) UnhandledPromiseRejectionWarning: Error: ER_NOT_SUPPORTED_AUTH_MODE: Client does not support authentication protocol requested by server; consider upgrading MySQL client
    at Handshake.Sequence._packetToError (/home/test/MapJS/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
    at Handshake.ErrorPacket (/home/test/MapJS/node_modules/mysql/lib/protocol/sequences/Handshake.js:123:18)
    at Protocol._parsePacket (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:291:23)
    at Parser._parsePacket (/home/test/MapJS/node_modules/mysql/lib/protocol/Parser.js:433:10)
    at Parser.write (/home/test/MapJS/node_modules/mysql/lib/protocol/Parser.js:43:10)
    at Protocol.write (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:38:16)
    at Socket.<anonymous> (/home/test/MapJS/node_modules/mysql/lib/Connection.js:88:28)
    at Socket.<anonymous> (/home/test/MapJS/node_modules/mysql/lib/Connection.js:526:10)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:295:12)
    --------------------
    at Protocol._enqueue (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:144:48)
    at Protocol.handshake (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:51:23)
    at Connection.connect (/home/test/MapJS/node_modules/mysql/lib/Connection.js:116:18)
    at getConnection (/home/test/MapJS/src/services/mysql.js:16:10)
    at /home/test/MapJS/src/services/mysql.js:38:26
    at new Promise (<anonymous>)
    at MySQLConnector.query (/home/test/MapJS/src/services/mysql.js:35:16)
    at isValidSession (/home/test/MapJS/src/services/session-store.js:38:28)
    at /home/test/MapJS/src/index.js:127:25
    at Layer.handle [as handle_request] (/home/test/MapJS/node_modules/express/lib/router/layer.js:95:5)
(node:44889) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:44889) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

MySQL 8.0.21. Config/db works with current master.

**Comment by versx** [comment]
> ```
> Failed to connect to the database: undefined
> (node:44889) UnhandledPromiseRejectionWarning: Error: ER_NOT_SUPPORTED_AUTH_MODE: Client does not support authentication protocol requested by server; consider upgrading MySQL client
>     at Handshake.Sequence._packetToError (/home/test/MapJS/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
>     at Handshake.ErrorPacket (/home/test/MapJS/node_modules/mysql/lib/protocol/sequences/Handshake.js:123:18)
>     at Protocol._parsePacket (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:291:23)
>     at Parser._parsePacket (/home/test/MapJS/node_modules/mysql/lib/protocol/Parser.js:433:10)
>     at Parser.write (/home/test/MapJS/node_modules/mysql/lib/protocol/Parser.js:43:10)
>     at Protocol.write (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:38:16)
>     at Socket.<anonymous> (/home/test/MapJS/node_modules/mysql/lib/Connection.js:88:28)
>     at Socket.<anonymous> (/home/test/MapJS/node_modules/mysql/lib/Connection.js:526:10)
>     at Socket.emit (events.js:315:20)
>     at addChunk (_stream_readable.js:295:12)
>     --------------------
>     at Protocol._enqueue (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:144:48)
>     at Protocol.handshake (/home/test/MapJS/node_modules/mysql/lib/protocol/Protocol.js:51:23)
>     at Connection.connect (/home/test/MapJS/node_modules/mysql/lib/Connection.js:116:18)
>     at getConnection (/home/test/MapJS/src/services/mysql.js:16:10)
>     at /home/test/MapJS/src/services/mysql.js:38:26
>     at new Promise (<anonymous>)
>     at MySQLConnector.query (/home/test/MapJS/src/services/mysql.js:35:16)
>     at isValidSession (/home/test/MapJS/src/services/session-store.js:38:28)
>     at /home/test/MapJS/src/index.js:127:25
>     at Layer.handle [as handle_request] (/home/test/MapJS/node_modules/express/lib/router/layer.js:95:5)
> (node:44889) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> (node:44889) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> ```
> 
> MySQL 8.0.21. Config/db works with current master.

Should be fixed now, damn typo lol.

**Comment by sketchysocks** [comment]
Hmm with ```allowMultipleSessions``` set to ```false``` it still let me login on my desktop and phone and use the map normally without any issues, unless I'm misunderstanding the purpose of this pr. I did see it clear some sessions upon reboot, but I'm not rebooting the front ends very often and it was easy to log back in on both anyway.

**Comment by versx** [comment]
> Hmm with `allowMultipleSessions` set to `false` it still let me login on my desktop and phone and use the map normally without any issues, unless I'm misunderstanding the purpose of this pr. I did see it clear some sessions upon reboot, but I'm not rebooting the front ends very often and it was easy to log back in on both anyway.

Hmm okay I'll need to do some more testing again later today. Basically if someone logs in and there are existing sessions for that same user ID in the `sessions` table it will delete the other sessions, which _should_ force the others to log out.

**Comment by sketchysocks** [comment]
Okay, earlier I had only tried refreshing and closing/reopening the tab on both devices simultaneously and wasn't seeing either being logged out. Just did some further testing and if the app is closed entirely, then it logs out.

- On desktop Chrome, it doesn't seem to be logging me out, even if I kill Chrome. 
- On mobile Chrome, I was getting logged out every time I closed the browser.

- Tried setting a new sessionSecret in config, logging in on my phone first, then the desktop, but any combination of closing/refreshing was always resulting in the desktop staying logged in and the mobile being logged out. 
- Same result for a new secret and never logging into it on my desktop, but still always being logged out on mobile every time I reopen Chrome. 
- Even tried clearing the session table but still the same results from both devices.

So I'm guessing something with the sql needs to be tweaked to "automatically" clear the other sessions, though that doesn't entirely explain why I was continuously logged out on mobile Chrome even when there was only 1 session with my user ID in the db. 

...if possible, would also like to a have config option with a list of user IDs to exclude from this, so I can add myself 👀

**Comment by versx** [comment]
> Okay, earlier I had only tried refreshing and closing/reopening the tab on both devices simultaneously and wasn't seeing either being logged out. Just did some further testing and if the app is closed entirely, then it logs out.
> 
> * On desktop Chrome, it doesn't seem to be logging me out, even if I kill Chrome.
> * On mobile Chrome, I was getting logged out every time I closed the browser.
> * Tried setting a new sessionSecret in config, logging in on my phone first, then the desktop, but any combination of closing/refreshing was always resulting in the desktop staying logged in and the mobile being logged out.
> * Same result for a new secret and never logging into it on my desktop, but still always being logged out on mobile every time I reopen Chrome.
> * Even tried clearing the session table but still the same results from both devices.
> 
> So I'm guessing something with the sql needs to be tweaked to "automatically" clear the other sessions, though that doesn't entirely explain why I was continuously logged out on mobile Chrome even when there was only 1 session with my user ID in the db.
> 
> ...if possible, would also like to a have config option with a list of user IDs to exclude from this, so I can add myself 👀

Thank you for the in-depth report!

File: src/index.js [hunk 1]
```diff
@@ -107,6 +121,14 @@ app.use(async (req, res, next) => {
         if (!config.discord.enabled) {
             return next();
         }
+        if (config.allowMultipleSessions) {
```
Comment (LordJavi) [inline src/index.js:124]:
```suggestion
        if (!config.allowMultipleSessions) {
```

**Comment by versx** [comment]
Should be working well now

**Pull request merged**